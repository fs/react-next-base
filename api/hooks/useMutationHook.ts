import type { DocumentNode } from 'graphql';
import type { TypedDocumentNode } from '@graphql-typed-document-node/core';
import type {
  MutationTuple,
  MutationHookOptions,
  FetchResult,
  MutationOptions,
  ApolloClient,
  InMemoryCache,
} from '@apollo/client';

import { useMutation } from '@apollo/client';

export const mutationApi = <TData, TVariables>(
  apolloClient: ApolloClient<InMemoryCache>,
  options: MutationOptions<TData, TVariables>,
): Promise<FetchResult<TData>> => {
  return apolloClient.mutate(options);
};

const useMutationHook = <TData, TVariables>(
  mutation: DocumentNode | TypedDocumentNode<TData, TVariables>,
  options?: MutationHookOptions<TData, TVariables> | undefined,
): MutationTuple<TData, TVariables> => {
  return useMutation(mutation, options);
};

export default useMutationHook;
