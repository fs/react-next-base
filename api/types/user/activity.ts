import User from 'domain/User';
import { PageInfo } from 'types/activityType';

export type Activity = {
  body: string;
  createdAt: string | Date;
  event: string;
  id: string | number;
  title: string;
  user: User;
};

export type ActivityEdge = {
  cursor: string;
  node: Activity;
};

export type ActivityData = {
  activities: {
    edges: ActivityEdge[];
    pageInfo: PageInfo;
  };
};

export type ActivityVariables = {
  events: string[];
  last?: number;
  before?: string;
  first?: number;
  after?: string;
};
