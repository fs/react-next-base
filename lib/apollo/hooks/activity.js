/* eslint-disable import/prefer-default-export */
import { useQuery } from '@apollo/client';
import Activities from 'graphql/queries/pages/activities.graphql';

import activityEvents from 'config/activityEvents';

const getFormattedActivity = data => {
  return data.activities.nodes.map(
    ({ id, title, body, createdAt, event, user: { firstName, lastName, email, avatarUrl } }) => ({
      id,
      title,
      description: body,
      date: new Date(createdAt).toLocaleString(),
      color: activityEvents.find(e => e.name === event).color,
      name: `${firstName} ${lastName}`,
      email,
      avatarUrl,
    }),
  );
};

export const useActivity = () => {
  const events = activityEvents.map(({ name }) => name);
  const { loading, error, data } = useQuery(Activities, { variables: { events } });

  return {
    activities: data ? getFormattedActivity(data) : null,
    loading,
    error,
  };
};
