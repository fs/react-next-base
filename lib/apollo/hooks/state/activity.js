/* eslint-disable import/prefer-default-export */
import { useQuery } from '@apollo/client';
import Activities from 'graphql/queries/pages/activities.graphql';

import activityEvents from 'config/activityEvents';
import activityPageSizes from 'config/activityPageSizes';

const getFormattedActivity = (data) => {
  return data.activities.edges.map(
    ({
      node: {
        id,
        title,
        body,
        createdAt,
        event,
        user: { firstName, lastName, email, avatarUrl },
      },
    }) => ({
      id,
      title,
      description: body,
      date: new Date(createdAt).toLocaleString(),
      color: activityEvents.find((e) => e.value === event).color,
      name: `${firstName} ${lastName}`,
      email,
      avatarUrl,
    }),
  );
};

export const useActivity = ({ beforeCursor, afterCursor, filterValue, pageSize = activityPageSizes[0] }) => {
  const events = filterValue ? [filterValue] : activityEvents.map(({ value }) => value);
  const first = afterCursor || (!afterCursor && !beforeCursor) ? pageSize : undefined;
  const last = beforeCursor ? pageSize : undefined;

  const { loading, error, data } = useQuery(Activities, {
    fetchPolicy: 'cache-and-network',
    nextFetchPolicy: 'cache-first',
    variables: {
      events,
      last,
      before: beforeCursor,
      first,
      after: afterCursor,
    },
  });

  return {
    activities: data ? getFormattedActivity(data) : null,
    pageInfo: data ? data.activities.pageInfo : null,
    loading,
    error,
  };
};
