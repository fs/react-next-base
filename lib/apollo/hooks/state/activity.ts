import { useQuery } from '@apollo/client';
import Activities from 'graphql/queries/pages/activities.graphql';

import activityEvents from 'config/activityEvents';
import activityPageSizes from 'config/activityPageSizes';

import User from 'domain/User';
import { PageInfo, Activity } from 'interfaces/activityType';

type ActivityParams = {
  beforeCursor?: string;
  afterCursor?: string;
  filterValue?: string;
  pageSize: number;
};

type Edge = {
  cursor: string;
  node: {
    body: string;
    createdAt: string | Date;
    event: string;
    id: string | number;
    title: string;
    user: User;
  };
};

type ActivityQueryState = {
  activities: {
    edges: Edge[];
    pageInfo: PageInfo;
  };
};

type ActivityQueryVars = {
  events: string[];
  last?: number;
  before?: string;
  first?: number;
  after?: string;
};

const getFormattedActivity = (edges: Edge[]): Activity[] => {
  return edges.map(
    ({
      node: {
        id,
        title,
        body,
        createdAt,
        event,
        user: { firstName, lastName, email, avatarUrl },
      },
    }) => ({
      id,
      title,
      description: body,
      date: new Date(createdAt).toLocaleString(),
      color: activityEvents.find((e) => e.value === event)?.color,
      name: `${firstName} ${lastName}`,
      email,
      avatarUrl,
    }),
  );
};

export const useActivity = ({
  beforeCursor,
  afterCursor,
  filterValue,
  pageSize = activityPageSizes[0],
}: ActivityParams) => {
  const events = filterValue ? [filterValue] : activityEvents.map(({ value }) => value);
  const first = afterCursor || (!afterCursor && !beforeCursor) ? pageSize : undefined;
  const last = beforeCursor ? pageSize : undefined;

  const { loading, error, data } = useQuery<ActivityQueryState, ActivityQueryVars>(Activities, {
    fetchPolicy: 'cache-and-network',
    nextFetchPolicy: 'cache-first',
    variables: {
      events,
      last,
      before: beforeCursor,
      first,
      after: afterCursor,
    },
  });

  return {
    activities: data?.activities?.edges ? getFormattedActivity(data?.activities.edges) : null,
    pageInfo: data?.activities?.pageInfo || null,
    loading,
    error,
  };
};
