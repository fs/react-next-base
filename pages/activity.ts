export { default } from 'components/pages/activity';
