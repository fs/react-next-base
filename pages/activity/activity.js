import React from 'react';
import { useQuery } from '@apollo/client';

import WithAuth from 'lib/auth/withAuth';
import WithAuthSecurity from 'lib/auth/withAuthSecurity';
import { withApolloClient } from 'lib/withApolloClient';

import Activities from 'graphql/queries/pages/activities.graphql';

import ErrorDecorator from 'decorators/ErrorDecorator';
import ErrorMessage from 'components/atoms/ErrorMessage';
import DefaultTemplate from 'components/templates/DefaultTemplate';
import ActivityTable from 'components/organisms/ActivityTable';

const getFormattedActivity = data => {
  return data.activities.nodes.map(
    ({ id, title, body, createdAt, event, user: { firstName, lastName, email, avatarUrl } }) => ({
      id,
      title,
      description: body,
      date: createdAt,
      type: event,
      name: `${firstName} ${lastName}`,
      email,
      avatarUrl,
    }),
  );
};

const Activity = () => {
  const { loading, error, data } = useQuery(Activities);

  const errorMessage = error ? new ErrorDecorator(error).getMessages() : null;

  return (
    <DefaultTemplate>
      {loading && <h3>Loading...</h3>}
      {error && <ErrorMessage>{errorMessage}</ErrorMessage>}
      {!loading && !error && <ActivityTable data={getFormattedActivity(data)} />}
    </DefaultTemplate>
  );
};

export default withApolloClient(WithAuth(WithAuthSecurity(Activity)));
