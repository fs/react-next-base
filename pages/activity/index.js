import Activity from './activity';

export default Activity;
