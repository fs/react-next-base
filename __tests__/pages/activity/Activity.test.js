import React from 'react';
import { render, screen } from '@testing-library/react';
import 'jest-styled-components';

import { useActivity } from 'lib/apollo/hooks/activity';
import renderWithApolloClient from '__tests__/helpers/renderWithApolloClient';
import renderWithTheme from '__tests__/helpers/renderWithTheme';
import activitiesMock from '__tests__/mocks/activitiesMock';

import Activity from 'pages/activity';

jest.mock('components/templates/DefaultTemplate', () => 'div');
jest.mock('lib/apollo/hooks/activity.js');
jest.mock('lib/apollo/hooks/actions.js', () => ({
  useSignIn: jest.fn(() => []),
  useUpdateUser: jest.fn(() => []),
}));

describe('Activity page', () => {
  test('should render correctly', async () => {
    // Arrange
    const mockUseActivity = jest.fn(() => ({
      loading: undefined,
      error: undefined,
      activities: activitiesMock,
    }));
    useActivity.mockImplementation(mockUseActivity);

    // Act
    render(renderWithTheme(renderWithApolloClient(<Activity />)));
    const pageContent = await screen.getByTestId('test-activity-table');

    // Assert
    expect(pageContent).toMatchSnapshot();
  });

  test('should show loader while loading', async () => {
    // Arrange
    const mockUseActivity = jest.fn(() => ({
      loading: true,
      error: undefined,
      activities: undefined,
    }));
    useActivity.mockImplementation(mockUseActivity);

    // Act
    render(renderWithTheme(renderWithApolloClient(<Activity />)));
    const loadingContent = await screen.getByTestId('test-activity-loading');

    // Assert
    expect(loadingContent).toMatchSnapshot();
  });

  test('should show error on error', async () => {
    // Arrange
    const error = {
      graphQLErrors: [
        {
          message:
            'Variable $events of type [ActivityEvent!] was provided invalid value for 0 (Expected "USER_LOGGED_IN1" to be one of: USER_LOGGED_IN, USER_REGISTERED, USER_RESET_PASSWORD, RESET_PASSWORD_REQUESTED, USER_UPDATED)',
          locations: [{ line: 1, column: 18 }],
        },
      ],
    };
    const mockUseActivityErrorData = jest.fn(() => ({
      loading: false,
      error,
      activities: undefined,
    }));
    useActivity.mockImplementation(mockUseActivityErrorData);

    // Act
    render(renderWithTheme(renderWithApolloClient(<Activity />)));
    const errorContent = await screen.getByTestId('test-activity-error');

    // Assert
    expect(errorContent).toMatchSnapshot();
  });
});
