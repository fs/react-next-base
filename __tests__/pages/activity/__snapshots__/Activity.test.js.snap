// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Activity page should render correctly 1`] = `
.c5 {
  display: inline-block;
  width: 1.875rem;
  height: 1.875rem;
  font-size: 0;
  border-radius: 50%;
}

.c3 {
  border-bottom: 1px solid #e1e1e1;
  padding: 0.8rem 0.6rem;
  font-size: 0.8rem;
  line-height: 1.6;
  font-weight: 100;
}

.c3:first-child {
  padding-left: 0;
}

.c3:nth-child(4) {
  white-space: nowrap;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 {
  position: relative;
  border-spacing: 0;
  display: block;
  overflow-x: auto;
  text-align: left;
  width: 100%;
  color: #606c76;
}

.c2 {
  min-width: 5px;
  background-color: #2979ff;
}

.c7 {
  min-width: 5px;
  background-color: #00e676;
}

.c8 {
  min-width: 5px;
  background-color: #f50057;
}

.c1 {
  border-bottom: 1px solid #e1e1e1;
  padding: 0.8rem 0.6rem;
  font-size: 0.8rem;
  line-height: 1.6;
}

.c1:first-child {
  padding-left: 0;
}

.c6 {
  margin-left: 0.5rem;
}

@media (min-width:calc(992px + 0.02px)) {
  .c3 {
    padding: 1.2rem 1rem;
    font-size: 0.9rem;
  }
}

@media (min-width:calc(992px + 0.02px)) {
  .c0 {
    display: table;
    overflow-x: initial;
  }
}

@media (min-width:calc(992px + 0.02px)) {
  .c1 {
    padding: 1.2rem 1rem;
    font-size: 0.9rem;
  }
}

<table
  class="c0"
  data-testid="test-activity-table"
>
  <thead>
    <tr>
      <th
        class="c1"
        colspan="2"
      >
        Title
      </th>
      <th
        class="c1"
        colspan="1"
      >
        Description
      </th>
      <th
        class="c1"
        colspan="1"
      >
        Date
      </th>
      <th
        class="c1"
        colspan="1"
      >
        User
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        class="c2"
        color="#2979ff"
      />
      <td
        class="c3"
      >
        <div
          class="c4"
        >
          User Updated
        </div>
      </td>
      <td
        class="c3"
      >
        <div
          class="c4"
        >
          User updated with the next attributes: First name - 111test1, Last name - еуые, Email - tes1t@test.com
        </div>
      </td>
      <td
        class="c3"
      >
        <div
          class="c4"
        >
          04.08.2020, 13:38:51
        </div>
      </td>
      <td
        class="c3"
      >
        <div
          class="c4"
        >
          <img
            alt="Avatar"
            class="c5"
            src="/images/avatar-placeholder.png"
          />
          <span
            class="c6"
          >
            111test1 еуые
             (
            tes1t@test.com
            )
          </span>
        </div>
      </td>
    </tr>
    <tr>
      <td
        class="c7"
        color="#00e676"
      />
      <td
        class="c3"
      >
        <div
          class="c4"
        >
          User Logged In
        </div>
      </td>
      <td
        class="c3"
      >
        <div
          class="c4"
        >
          User logged in with the next attributes: First name - , Last name - 
        </div>
      </td>
      <td
        class="c3"
      >
        <div
          class="c4"
        >
          04.08.2020, 15:01:32
        </div>
      </td>
      <td
        class="c3"
      >
        <div
          class="c4"
        >
          <img
            alt="Avatar"
            class="c5"
            src="/images/avatar-placeholder.png"
          />
          <span
            class="c6"
          >
            Nikita Polyakov
             (
            nikita.polyakov@flatstack.com
            )
          </span>
        </div>
      </td>
    </tr>
    <tr>
      <td
        class="c8"
        color="#f50057"
      />
      <td
        class="c3"
      >
        <div
          class="c4"
        >
          User Registered
        </div>
      </td>
      <td
        class="c3"
      >
        <div
          class="c4"
        >
          New user registered with the next attributes: First name - Test, Last name - Test↵
        </div>
      </td>
      <td
        class="c3"
      >
        <div
          class="c4"
        >
          04.08.2020, 15:26:08
        </div>
      </td>
      <td
        class="c3"
      >
        <div
          class="c4"
        >
          <img
            alt="Avatar"
            class="c5"
            src="/images/avatar-placeholder.png"
          />
          <span
            class="c6"
          >
            Test Test
             (
            test123@example.com
            )
          </span>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Activity page should show error on error 1`] = `
<div
  data-testid="test-activity-error"
>
  <p>
    Something went wrong
  </p>
</div>
`;

exports[`Activity page should show loader while loading 1`] = `
<h3
  data-testid="test-activity-loading"
>
  Loading...
</h3>
`;
